Sming/Components/rboot/component.mk

```diff
@@ -21,37 +21,67 @@ endif
 # => APP
 
 # rBoot options, overwrite them in the projects Makefile-user.mk
-CONFIG_VARS				+= RBOOT_BIG_FLASH RBOOT_TWO_ROMS RBOOT_RTC_ENABLED RBOOT_GPIO_ENABLED RBOOT_GPIO_SKIP_ENABLED
-RBOOT_BIG_FLASH			?= 1
-RBOOT_TWO_ROMS			?= 0
+CONFIG_VARS				+= RBOOT_RTC_ENABLED RBOOT_GPIO_ENABLED RBOOT_GPIO_SKIP_ENABLED
 RBOOT_RTC_ENABLED		?= 0
 RBOOT_GPIO_ENABLED		?= 0
 # RBOOT_GPIO_SKIP_ENABLED and RBOOT_GPIO_ENABLED cannot be used at the same time.
 RBOOT_GPIO_SKIP_ENABLED	?= 0
 
 ifeq ($(RBOOT_GPIO_ENABLED)$(RBOOT_GPIO_SKIP_ENABLED),11)
-$(error "Cannot enable RBOOT_GPIO_ENABLED and RBOOT_GPIO_SKIP_ENABLED at the same time)
+$(error Cannot enable RBOOT_GPIO_ENABLED and RBOOT_GPIO_SKIP_ENABLED at the same time)
 endif
 
 ### ROM Addresses ###
-# The parameter below specifies the location of the second rom.
-# This parameter is used only when RBOOT_BIG_FLASH = 1 
-# BOOT_ROM1_ADDR = 0x200000
+# Make sure that your ROM slots and SPIFFS slot(s) do not overlap!
+
+CONFIG_VARS				+= RBOOT_ROM0_ADDR RBOOT_ROM1_ADDR RBOOT_ROM2_ADDR
+
+# Loation of first ROM (default is sector 2 after rboot and rboot config sector)
+RBOOT_ROM0_ADDR			?= 0x002000
+
+# The parameter below specifies the location of the second ROM.
+# You need a second slot for any kind of firmware update mechanism.
+# Leave empty if you don't need a second ROM slot.
+RBOOT_ROM1_ADDR			?=
 
 # The parameter below specifies the location of the GPIO ROM.
-# This parameter is used only when RBOOT_GPIO_ENABLED = 1
-# If you use two SPIFFS make sure that this address is minimum
-# RBOOT_SPIFFS_1 + SPIFF_SIZE 
-# BOOT_ROM2_ADDR = 0x310000
+# It is only used when RBOOT_GPIO_ENABLED = 1
+# Note that setting this parameter will only configure rboot.
+# The Sming build system does not create a ROM for this slot.
+RBOOT_ROM2_ADDR			?=
+
+ifeq ($(RBOOT_GPIO_ENABLED),0)
+ifneq ($(RBOOT_ROM2_ADDR),)
+$(warning RBOOT_GPIO_ENABLED is 0, RBOOT_ROM2_ADDR will be ignored)
+RBOOT_ROM2_ADDR :=
+endif
+endif
+
+# determine number of roms to generate
+ifneq ($(RBOOT_ROM1_ADDR),)
+RBOOT_TWO_ROMS := $(shell awk 'BEGIN { print ($(RBOOT_ROM0_ADDR) % 0x100000) == ($(RBOOT_ROM1_ADDR) % 0x100000) ? 0 : 1 }')
+else
+RBOOT_TWO_ROMS := 0
+endif
+
+DEBUG_VARS 				+= RBOOT_TWO_ROMS
+
+# BIGFLASH mode is needed if at least one ROM address exceeds the first 1MB of flash
+BIGFLASH_TEST := awk 'BEGIN { big=0; for(i = 1; i < ARGC; ++i) if(strtonum(ARGV[i]) > 0x100000) big=1; print big }'
```

Comment:
@aemseemann I am getting an error on my Linux box:

```
awk: line 2: function strtonum never defined
```

Reply:
@mikee47  Can you suggest a simple check that should replace the awk code and work on Windows, Mac OS X, Linux (and optionally FreeBSD) ?

Reply:
I've seen `int()` used, but having other issues getting this to run. Line 178...

Reply:
As I recall we had issues getting compatibility with mawk, I think if the standard were gawk we'd be fine.

Reply:
Wasn't aware of the different flavours of awk... I'll try to get it to work with mawk and busybox awk. Hopefully the latter will be minimal enough to cover all forseable installations.

Reply:
This is more complicated than I thought. It seems I can get it to work for `mawk '...'` and `gawk -n '...'` (or `awk -n` if that is an alias), but have no idea how to detect which one to use.
Since this is quite a core element of the build process, forcing the user to choose via `make AWK=...` is not acceptable either.

As alternatives to awk, how do you feel about...
- `$(( ... ))`  (or, more generally, which shells have to be supported?)
- `python -c '...'`

Personally, I would prefer the Python option, since it is probably the most platform independent solution and Python is already used by the build system.

Reply:
The changes work on my Linux box. @mikee47 When you have time can you check if that works on your Windows box? I will check later today also the MacOS build.

Reply:
In MinGW, which uses gawk, it's not picking up the rom_base or rom_offset variables:

```
irom0_0_seg :                         org = 0x00000000, len = 1M - 0x10
```

According to https://web.mit.edu/gnu/doc/html/gawk_16.html variables are available within a BEGIN rule, but says nothing about their availablity within a match. I'll post back if I can find a working solution.

Reply:
Tried a variety different approaches and it's just not happy converting strings. Rather than attempting to filter `rboot_rom0.ld`, would it simplify things if we just generate the file entirely, or generated the required fragment and include it? Could we use the compiler to convert .h to .ld ?

Reply:
I'm mindful that at present the build system doen't provide any way for Components to customise the linker scripts, something to look at later I think.

Reply:
Next attempt: I now moved all the math to the linker script itself, so awk does not have to do hex conversion and variables anymore. Should look like this for the first rom:
```
  irom0_0_seg : org = 0x40200010 + (0x2000 & 0xFFFFF), len = 1M - (0x2000 & 0xFFFFF) - 0x10
```

@mikee47, @slaff: Can you try again, please?

Reply:
Works a treat!

Reply:
@mikee47 I am waiting for your approval to schedule this PR for merging?